zerver/tests/test_auth_backends.py

```diff
@@ -739,7 +739,12 @@ def prepare_login_url_and_headers(
         params = {}
         headers = {}
         if subdomain is not None:
-            headers['HTTP_HOST'] = subdomain + ".testserver"
+            if subdomain == '':
+                # "testserver" may trip up some libraries' URL validation,
```

Comment:
Can you explain the details a bit more here?  I'd like to know what library, etc.

Reply:
Also, do we want to be encoding the root domain both with subdomain=None and subdomain=''?  Would it be cleaner to just have test_social_auth_saml_idp_org_membership_root_subdomain pass `subdomain='www'`?

Reply:
> Would it be cleaner to just have test_social_auth_saml_idp_org_membership_root_subdomain pass subdomain='www'?

Hmm, it crossed my mind, but I think we should have the `if subdomain ==''` case, because `subdomain + ".testserver"` is just invalid in the case of `''` subdomain. So might as well have this block the way I did it :thinking: One the other hand `subdomain=None` case didn't really make sense and wasn't used (except one test that simply forgot to specify this kwarg), so I pushed a commit to clean that up.

Reply:
> Can you explain the details a bit more here? I'd like to know what library, etc.

python3-saml has this snippet that doesn't like the `testserver` url https://github.com/onelogin/python3-saml/blob/master/src/onelogin/saml2/settings.py#L32
```
# Regex from Django Software Foundation and individual contributors.
# Released under a BSD 3-Clause License
url_regex = re.compile(
    r'^(?:[a-z0-9\.\-]*)://'  # scheme is validated separately
    r'(?:(?:[A-Z0-9_](?:[A-Z0-9-_]{0,61}[A-Z0-9_])?\.)+(?:[A-Z]{2,6}\.?|[A-Z0-9-]{2,}\.?)|'  # domain...
    r'localhost|'  # localhost...
    r'\d{1,3}\.\d{1,3}\.\d{1,3}\.\d{1,3}|'  # ...or ipv4
    r'\[?[A-F0-9]*:[A-F0-9:]+\]?)'  # ...or ipv6
    r'(?::\d+)?'  # optional port
    r'(?:/?|[/?]\S+)$', re.IGNORECASE)
url_schemes = ['http', 'https', 'ftp', 'ftps']


def validate_url(url):
    """
    Auxiliary method to validate an urllib
    :param url: An url to be validated
    :type url: string
    :returns: True if the url is valid
    :rtype: bool
    """

    scheme = url.split('://')[0].lower()
    if scheme not in url_schemes:
        return False
    if not bool(url_regex.search(url)):
        return False
    return True
```